File: fixtures/scale/dart/medium/lib/history/history.dart
import 'history_entry.dart';

class History {
  final List<HistoryEntry> _entries = [];
  final int capacity;

  History(this.capacity);

  void record(String input, double result) {
    if (_entries.length == capacity) _entries.removeAt(0);
    _entries.add(HistoryEntry(input, result));
  }

  List<HistoryEntry> recent(int count) {
    final from = (_entries.length - count).clamp(0, _entries.length);
    return _entries.sublist(from);
  }

  int size() => _entries.length;
}

File: fixtures/scale/scala/medium/src/calc/history/History.scala
package calc.history

import scala.collection.mutable.ArrayDeque

final class History(capacity: Int) {
  private val entries = ArrayDeque.empty[HistoryEntry]

  def record(input: String, result: Double): Unit = {
    if (entries.size == capacity) entries.removeHead()
    entries.append(HistoryEntry.now(input, result))
  }

  def recent(count: Int): List[HistoryEntry] =
    entries.takeRight(count).toList

  def size: Int = entries.size
}
